**1. Import numpy and pandas libraries**

In [3]:
import numpy as np
import pandas as pd

In [4]:
#Test the libraries
print("Numpy version: ", np.__version__)
print("Pandas version: ", pd.__version__)

Numpy version:  1.23.4
Pandas version:  1.5.1


**2. Load datasets**

In [5]:
# Read the data
Books = pd.read_csv('dataset/books_clean_dataset.csv', on_bad_lines='skip', sep=';', encoding='latin-1')
Users = pd.read_csv('dataset/bx_278860_users.csv', on_bad_lines='skip', sep=';', encoding='latin-1')
Ratings = pd.read_csv('dataset/bx_1149781_book_ratings.csv', on_bad_lines='skip', sep=';', encoding='latin-1')

C:\Users\almud\AppData\Local\Temp\ipykernel_12260\1522618593.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  Books = pd.read_csv('dataset/bx_271380_books.csv', on_bad_lines='skip', sep=';', encoding='latin-1')


I have got 'Python Pandas Error tokenizing data' before adding the following code:
**on_bad_lines='skip', sep=';', encoding='latin-1**


In [6]:
Books.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [7]:
Users.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [8]:
Ratings.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [9]:
Books.shape

(271360, 8)

In [10]:
Ratings.shape

(1149780, 3)

In [11]:
Users.shape

(278858, 3)

## Find out the missing variables

In [12]:
Books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [13]:
Ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [14]:
Users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

## Find out the duplication(s)

In [15]:
Books.duplicated().sum()

0

In [16]:
Ratings.duplicated().sum()

0

In [17]:
Users.duplicated().sum()

0

## Popularity-Based Filtering Recommendation System:

**Here we will display the top N highest rated books ( N = number of books)**

In [18]:
Ratings

User-ID         ISBN  Book-Rating
0         276725   034545104X            0
1         276726   0155061224            5
2         276727   0446520802            0
3         276729   052165615X            3
4         276729   0521795028            6
...          ...          ...          ...
1149775   276704   1563526298            9
1149776   276706   0679447156            0
1149777   276709   0515107662           10
1149778   276721   0590442449           10
1149779   276723  05162443314            8

[1149780 rows x 3 columns]

In [19]:
Books

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

In [20]:
name_ratings = pd.merge(Books, Ratings, on='ISBN')
number_ratings_df = name_ratings.groupby('Book-Title').count()['Book-Rating'].reset_index()
number_ratings_df.rename(columns={'Book-Rating': 'Number-of-Ratings'}, inplace=True)

In [21]:
number_ratings_df

Book-Title  Number-of-Ratings
0        A Light in the Storm: The Civil War Diary of ...                  4
1                                   Always Have Popsicles                  1
2                    Apple Magic (The Collector's series)                  1
3        Ask Lily (Young Women of Faith: Lily Series, ...                  1
4        Beyond IBM: Leadership Marketing and Finance ...                  1
...                                                   ...                ...
241066                                  �?�?lpiraten.                  2
241067                 �?�?rger mit Produkt X. Roman.                  4
241068                            �?�?sterlich leben.                  1
241069                          �?�?stlich der Berge.                  3
241070                              �?�?thique en toc                  2

[241071 rows x 2 columns]

In [22]:
average_rating_df = name_ratings.groupby('Book-Title').mean()['Book-Rating'].reset_index()
average_rating_df.rename(columns={'Book-Rating': 'Average-Rating'}, inplace=True)

C:\Users\almud\AppData\Local\Temp\ipykernel_12260\3743345720.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  average_rating_df = name_ratings.groupby('Book-Title').mean()['Book-Rating'].reset_index()


In [23]:
average_rating_df

Book-Title  Average-Rating
0        A Light in the Storm: The Civil War Diary of ...        2.250000
1                                   Always Have Popsicles        0.000000
2                    Apple Magic (The Collector's series)        0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...        8.000000
4        Beyond IBM: Leadership Marketing and Finance ...        0.000000
...                                                   ...             ...
241066                                  �?�?lpiraten.        0.000000
241067                 �?�?rger mit Produkt X. Roman.        5.250000
241068                            �?�?sterlich leben.        7.000000
241069                          �?�?stlich der Berge.        2.666667
241070                              �?�?thique en toc        4.000000

[241071 rows x 2 columns]

In [24]:
popularity_df = pd.merge(number_ratings_df, average_rating_df, on='Book-Title')

In [25]:
popularity_df

Book-Title  Number-of-Ratings  \
0        A Light in the Storm: The Civil War Diary of ...                  4   
1                                   Always Have Popsicles                  1   
2                    Apple Magic (The Collector's series)                  1   
3        Ask Lily (Young Women of Faith: Lily Series, ...                  1   
4        Beyond IBM: Leadership Marketing and Finance ...                  1   
...                                                   ...                ...   
241066                                  �?�?lpiraten.                  2   
241067                 �?�?rger mit Produkt X. Roman.                  4   
241068                            �?�?sterlich leben.                  1   
241069                          �?�?stlich der Berge.                  3   
241070                              �?�?thique en toc                  2   

        Average-Rating  
0             2.250000  
1             0.000000  
2             0.000000  
3             8.000000  
4             0.000000  
...                ...  
241066        0.000000  
241067        5.250000  
241068        7.000000  
241069        2.666667  
241070        4.000000  

[241071 rows x 3 columns]

In [26]:
popularity_df.sort_values('Average-Rating', ascending=False)

Book-Title  Number-of-Ratings  \
66223              Film Is: The International Free Cinema                  1   
123125  More Secrets of Happy Children: Embrace Your P...                  1   
96558             Jo's Boys : From the Original Publisher                  1   
212107    The Vanished Priestess : An Annie Szabo Mystery                  1   
72503                                    Game and Hunting                  1   
...                                                   ...                ...   
93628   Invaders of the Planet Earth (Choose Your Own ...                  1   
154021    Sammy Discovers the Alphabet (Learn-With-Sammy)                  1   
154019                    Sammy Carducci's Guide to Women                  2   
154018  Sammle mir Kiesel am Flu�?�?. Mehr als ein...                  1   
171214                                     TRAVELLING MAN                  1   

        Average-Rating  
66223             10.0  
123125            10.0  
96558             10.0  
212107            10.0  
72503             10.0  
...                ...  
93628              0.0  
154021             0.0  
154019             0.0  
154018             0.0  
171214             0.0  

[241071 rows x 3 columns]

**Take the books that have rated 250 or more**

In [36]:
popularity_df = popularity_df[popularity_df['Number-of-Ratings']>=250]

In [28]:
popularity_df

Book-Title  Number-of-Ratings  \
764                                                  1984                284   
818                                   1st to Die: A Novel                509   
1048                                           2nd Chance                356   
1760                                   A Bend in the Road                346   
2281    A Child Called \It\": One Child's Courage to S...                265   
...                                                   ...                ...   
233850                           White Oleander : A Novel                387   
233851       White Oleander : A Novel (Oprah's Book Club)                356   
234740  Wicked: The Life and Times of the Wicked Witch...                326   
234951                                        Wild Animus               2502   
240860                                \O\" Is for Outlaw"                269   

        Average-Rating  
764           4.454225  
818           3.575639  
1048          3.269663  
1760          3.364162  
2281          4.086792  
...                ...  
233850        3.506460  
233851        3.772472  
234740        3.766871  
234951        1.019584  
240860        2.988848  

[186 rows x 3 columns]

In [34]:
popularity_df = popularity_df.sort_values('Average-Rating', ascending=False)

In [35]:
popularity_df

Book-Title  Number-of-Ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)                428   
80422        Harry Potter and the Goblet of Fire (Book 4)                387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)                278   
80426   Harry Potter and the Order of the Phoenix (Boo...                347   
80414    Harry Potter and the Chamber of Secrets (Book 2)                556   
...                                                   ...                ...   
227701       Vinegar Hill (Oprah's Book Club (Paperback))                265   
233635                                           Whispers                286   
143377                                  Presumed Innocent                294   
94382                                        Isle of Dogs                288   
234951                                        Wild Animus               2502   

        Average-Rating  
80434         5.852804  
80422         5.824289  
80441         5.737410  
80426         5.501441  
80414         5.183453  
...                ...  
227701        2.245283  
233635        2.199301  
143377        2.139456  
94382         2.000000  
234951        1.019584  

[186 rows x 3 columns]

In [37]:
popularity_df

Book-Title  Number-of-Ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)                428   
80422        Harry Potter and the Goblet of Fire (Book 4)                387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)                278   
80426   Harry Potter and the Order of the Phoenix (Boo...                347   
80414    Harry Potter and the Chamber of Secrets (Book 2)                556   
...                                                   ...                ...   
227701       Vinegar Hill (Oprah's Book Club (Paperback))                265   
233635                                           Whispers                286   
143377                                  Presumed Innocent                294   
94382                                        Isle of Dogs                288   
234951                                        Wild Animus               2502   

        Average-Rating  
80434         5.852804  
80422         5.824289  
80441         5.737410  
80426         5.501441  
80414         5.183453  
...                ...  
227701        2.245283  
233635        2.199301  
143377        2.139456  
94382         2.000000  
234951        1.019584  

[186 rows x 3 columns]

**Display the full info for the most popular books**

In [38]:
popularity_df = popularity_df.merge(Books, on='Book-Title').drop_duplicates('Book-Title')

In [39]:
popularity_df

Book-Title  Number-of-Ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)                428   
3         Harry Potter and the Goblet of Fire (Book 4)                387   
5       Harry Potter and the Sorcerer's Stone (Book 1)                278   
9    Harry Potter and the Order of the Phoenix (Boo...                347   
13    Harry Potter and the Chamber of Secrets (Book 2)                556   
..                                                 ...                ...   
716       Vinegar Hill (Oprah's Book Club (Paperback))                265   
717                                           Whispers                286   
727                                  Presumed Innocent                294   
733                                       Isle of Dogs                288   
739                                        Wild Animus               2502   

     Average-Rating        ISBN        Book-Author Year-Of-Publication  \
0          5.852804  0439136350      J. K. Rowling                1999   
3          5.824289  0439139597      J. K. Rowling                2000   
5          5.737410  0590353403      J. K. Rowling                1998   
9          5.501441  043935806X      J. K. Rowling                2003   
13         5.183453  0439064872      J. K. Rowling                2000   
..              ...         ...                ...                 ...   
716        2.245283  0380730138   A. Manette Ansay                1998   
717        2.199301  0440216745        BELVA PLAIN                1994   
727        2.139456  0446359866        Scott Turow                1989   
733        2.000000  0425182908  Patricia Cornwell                2002   
739        1.019584  0971880107       Rich Shapero                2004   

                    Publisher  \
0                  Scholastic   
3                  Scholastic   
5                  Scholastic   
9                  Scholastic   
13                 Scholastic   
..                        ...   
716                 Perennial   
717                      Dell   
727              Warner Books   
733  Berkley Publishing Group   
739                   Too Far   

                                           Image-URL-S  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
5    http://images.amazon.com/images/P/0590353403.0...  
9    http://images.amazon.com/images/P/043935806X.0...  
13   http://images.amazon.com/images/P/0439064872.0...  
..                                                 ...  
716  http://images.amazon.com/images/P/0

**The variables that we are going to use in the recommendation:**

In [31]:
#ISBN,
# Book-Title, 
# Book-Author,
# Number-of-Ratings,
# Average-Rating,
# Image-URL-S, 
# Image-URL-M, 
# Image-URL-L

In [40]:
popularity_df = popularity_df[['ISBN','Book-Title', 'Book-Author', 'Number-of-Ratings', 'Average-Rating', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L']]

In [41]:
popularity_df

ISBN                                         Book-Title  \
0    0439136350  Harry Potter and the Prisoner of Azkaban (Book 3)   
3    0439139597       Harry Potter and the Goblet of Fire (Book 4)   
5    0590353403     Harry Potter and the Sorcerer's Stone (Book 1)   
9    043935806X  Harry Potter and the Order of the Phoenix (Boo...   
13   0439064872   Harry Potter and the Chamber of Secrets (Book 2)   
..          ...                                                ...   
716  0380730138       Vinegar Hill (Oprah's Book Club (Paperback))   
717  0440216745                                           Whispers   
727  0446359866                                  Presumed Innocent   
733  0425182908                                       Isle of Dogs   
739  0971880107                                        Wild Animus   

           Book-Author  Number-of-Ratings  Average-Rating  \
0        J. K. Rowling                428        5.852804   
3        J. K. Rowling                387        5.824289   
5        J. K. Rowling                278        5.737410   
9        J. K. Rowling                347        5.501441   
13       J. K. Rowling                556        5.183453   
..                 ...                ...             ...   
716   A. Manette Ansay                265        2.245283   
717        BELVA PLAIN                286        2.199301   
727        Scott Turow                294        2.139456   
733  Patricia Cornwell                288        2.000000   
739       Rich Shapero               2502        1.019584   

                                           Image-URL-S  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
5    http://images.amazon.com/images/P/0590353403.0...   
9    http://images.amazon.com/images/P/043935806X.0...   
13   http://images.amazon.com/images/P/0439064872.0...   
..                                                 ...   
716  http://images.amazon.com/images/P/0380730138.0...   
717  http://images.amazon.com/images/P/0440216745.0...   
727  http://images.amazon.com/images/P/0446359866.0...   
733  http://images.amazon.com/images/P/0425182908.0...   
739  http://images.amazon.com/images/P/0971880107.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
5    http://images.amazon.com/images/P/0590353403.0...  
9    http://images.amazon.com/images/P/043935806X.0...  
13   http://images.amazon.com/images/P/0439064872.0...  
..                                                 ...  
716  http://images.amazon.com/images/P/0380730138.0...  
717  http://images.amazon.com/images/P/0440216745.0...  
727  http://images.amazon.com/images/P/0446359866.0...  
733  http://images.amazon.com/images/P/0425182908.0...  
739  http://images.amazon.com/images/P/0971880107.0...  

[186 rows x 8 columns]

### We have above the most popular 186 books based on popularity

In [43]:
def recommend_most_popular_books(num):
    return popularity_df.head(num)

In [44]:
#Let's get the first 10 highest rated books with at least 250 ratings
recommend_most_popular_books(10)

ISBN                                         Book-Title  \
0   0439136350  Harry Potter and the Prisoner of Azkaban (Book 3)   
3   0439139597       Harry Potter and the Goblet of Fire (Book 4)   
5   0590353403     Harry Potter and the Sorcerer's Stone (Book 1)   
9   043935806X  Harry Potter and the Order of the Phoenix (Boo...   
13  0439064872   Harry Potter and the Chamber of Secrets (Book 2)   
16  0345339681  The Hobbit : The Enchanting Prelude to The Lor...   
17  0345339703  The Fellowship of the Ring (The Lord of the Ri...   
26  059035342X  Harry Potter and the Sorcerer's Stone (Harry P...   
28  0345339711     The Two Towers (The Lord of the Rings, Part 2)   
39  0446310786                              To Kill a Mockingbird   

       Book-Author  Number-of-Ratings  Average-Rating  \
0    J. K. Rowling                428        5.852804   
3    J. K. Rowling                387        5.824289   
5    J. K. Rowling                278        5.737410   
9    J. K. Rowling                347        5.501441   
13   J. K. Rowling                556        5.183453   
16  J.R.R. TOLKIEN                281        5.007117   
17  J.R.R. TOLKIEN                368        4.948370   
26   J. K. Rowling                575        4.895652   
28  J.R.R. TOLKIEN                260        4.880769   
39      Harper Lee                510        4.700000   

                                          Image-URL-S  \
0   http://images.amazon.com/images/P/0439136350.0...   
3   http://images.amazon.com/images/P/0439139597.0...   
5   http://images.amazon.com/images/P/0590353403.0...   
9   http://images.amazon.com/images/P/043935806X.0...   
13  http://images.amazon.com/images/P/0439064872.0...   
16  http://images.amazon.com/images/P/0345339681.0...   
17  http://images.amazon.com/images/P/0345339703.0...   
26  http://images.amazon.com/images/P/059035342X.0...   
28  http://images.amazon.com/images/P/0345339711.0...   
39  http://images.amazon.com/images/P/0446310786.0...   

                                          Image-URL-M  \
0   http://images.amazon.com/images/P/0439136350.0...   
3   http://images.amazon.com/images/P/0439139597.0...   
5   http://images.amazon.com/images/P/0590353403.0...   
9   http://images.amazon.com/images/P/043935806X.0...   
13  http://images.amazon.com/images/P/0439064872.0...   
16  http://images.amazon.com/images/P/0345339681.0...   
17  http://images.amazon.com/images/P/0345339703.0...   
26  http://images.amazon.com/images/P/059035342X.0...   
28  http://images.amazon.com/images/P/0345339711.0...   
39  http://images.amazon.com/images/P/0446310786.0...   

                                          Image-URL-L  
0   http://images.amazon.com/images/P/0439136350.0...  
3   http://images.amazon.com/images/P/0439139597.0...  
5   http://images.amazon.com/images/P/0590353403.0...  
9   http://images.amazon.com/images/P/043935806X.0...  
13  http://images.amazon.com/images/P/0439064872.0...  
16  http://images.amazon.com/images/P/0345339681.0...  
17  http://images.amazon.com/images/P/0345339703.0...  
26  http://images.amazon.com/images/P/059035342X.0...  
28  http://images.amazon.com/images/P/0345339711.0...  
39  http://images.amazon.com/images/P/0446310786.0...